ozaria/site/components/interactive/PageInteractiveEditor/index.vue

```diff
@@ -0,0 +1,208 @@
+
+require('lib/setupTreema')
+
+module.exports = Vue.extend({
+  props: {
+    slug: String
+  },
+  data: () => ({
+    interactive: null,
+    treema: null,
+    interactiveList: null,
+    valid: true,
+    state: {
+      saving: false
+    },
+    interactiveSlug: ""
+  }),
+  components: {
+    'list-item': ListItem
+  },
+  async created () {
+    if (!me.hasInteractiveAccess()) {
+      alert('You must be logged in as an admin to use this page.')
+      return application.router.navigate('/editor', { trigger: true })
+    }
+    console.log(`Got the slug: ${this.slug}`)
+    this.interactiveSlug = this.slug
+    if (this.interactiveSlug)  {
+      await this.fetchInteractive(this.interactiveSlug)
+    }
+    else {
+      await this.fetchList()
+    }
+  },
+  methods: {
+    /**
+     * Fetch and populate treema with interactive slug.
+     * Clears the list
+     */
+    async fetchInteractive(slug) {
+      this.interactiveList = null
+      this.interactiveSlug = slug
+      try {
+        this.interactive = new Interactive(await getInteractive(slug))
+      } catch (e) {
+        noty({ text: `Error finding slug '${slug}'.`, type: 'error', timeout: '1000' })
+        return this.fetchList()
+      }
+      const data = $.extend(true, {}, this.interactive.attributes)
+      const el = $(`<div></div>`);
+      const treemaOptions = {
+        data: data,
+        schema: Interactive.schema,
+        callbacks: {
+          change: this.onTreemaChanged
+        }
+      }
+      const treema = this.treema = TreemaNode.make(el, treemaOptions)
+      treema.build()
+      $(this.$refs.treemaEditor).append(el)
+    },
+
+    /**
+     * Fetch all names and slugs of interactives from the database.
+     * Clears the slug and treema.
+     */
+    async fetchList () {
+      this.interactive = null
+      this.interactiveSlug = ''
+      this.treema = null
+      $(this.$refs.treemaEditor).children().remove()
+      try {
+        this.interactiveList = await getAllInteractives();
+      } catch (e) {
+        console.error("Error while fetching the interactive list:", e)
+        noty({ text: 'Error occured while fetching the interactive list', type: 'error', timeout: 1000 })
+      }
+    },
+
+    /**
+     * Performs schema validation and pushes changes from treema to the interactive model.
+     */
+    onTreemaChanged() {
+      const ajv = new Ajv(ozariaUtils.getAjvOptions())
+      const data = this.treema.data
+      this.valid = ajv.validate(Interactive.schema, data)
+      if (this.valid) {
+        this.interactive.set(data)
+      } 
+      else {
+        console.error('Schema validation error', ajv.errors)
+        noty({
+          text: 'Schema validation error.',
+          type:'error',
+          timeout: 1000
+        })
+      }
+    },
+
+    /**
+     * Saves the properties of the interactive to the database.
+     */
+    async saveInteractive() {
+      if (this.valid) {
```

Comment:
Should we do another validation in case a change event didn't fire? The server does do another validation so it's fine.

Reply:
Sorry I didn't understand the question `Should we do another validation in case a change event didn't fire?`
Thanks for the nitpick.

Reply:
I think my question doesn't matter now that I have looked at other code